ksqldb-engine/src/main/java/io/confluent/ksql/query/PullQueryQueue.java

```diff
@@ -0,0 +1,197 @@
+
+  private final BlockingQueue<PullQueryRow> rowQueue;
+  private final int offerTimeoutMs;
+  private volatile boolean closed = false;
+
+  private LimitHandler limitHandler;
+  private LimitQueueCallback callback;
+
+  public PullQueryQueue() {
+    this(BLOCKING_QUEUE_CAPACITY, 100);
+  }
+
+  public PullQueryQueue(
+      final int queueSizeLimit,
+      final int offerTimeoutMs) {
+    this.callback = new QueueToCompletionCallback(this::isClosed);
+    this.rowQueue = new LinkedBlockingQueue<>(queueSizeLimit);
+    this.offerTimeoutMs = offerTimeoutMs;
+    this.limitHandler = () -> { };
+  }
+
+  @Override
+  public void setLimitHandler(final LimitHandler limitHandler) {
+    this.limitHandler = limitHandler;
+    callback.setLimitHandler(limitHandler);
+  }
+
+  @Override
+  public void setQueuedCallback(final Runnable queuedCallback) {
+    final LimitQueueCallback parent = callback;
+
+    callback = new LimitQueueCallback() {
+      @Override
+      public boolean shouldQueue() {
+        return parent.shouldQueue();
+      }
+
+      @Override
+      public void onQueued() {
+        parent.onQueued();
+        queuedCallback.run();
+      }
+
+      @Override
+      public void setLimitHandler(final LimitHandler limitHandler) {
+        parent.setLimitHandler(limitHandler);
+      }
+    };
+  }
+
+  @Override
+  public KeyValue<List<?>, GenericRow> poll(final long timeout, final TimeUnit unit)
+      throws InterruptedException {
+    return Optional.ofNullable(rowQueue.poll(timeout, unit))
+        .map(PQ_ROW_TO_KEY_VALUE)
+        .orElse(null);
+  }
+
+  @Override
+  public KeyValue<List<?>, GenericRow> poll() {
+    return Optional.ofNullable(rowQueue.poll())
+        .map(PQ_ROW_TO_KEY_VALUE)
+        .orElse(null);
+  }
+
+  @Override
+  public void drainTo(final Collection<? super KeyValue<List<?>, GenericRow>> collection) {
+    final List<PullQueryRow> list = new ArrayList<>();
+    drainRowsTo(list);
+    list.stream()
+        .map(PQ_ROW_TO_KEY_VALUE)
+        .forEach(collection::add);
+  }
+
+  public PullQueryRow pollRow(final long timeout, final TimeUnit unit) throws InterruptedException {
+    return rowQueue.poll(timeout, unit);
+  }
+
+  public void drainRowsTo(final Collection<PullQueryRow> collection) {
+    rowQueue.drainTo(collection);
+  }
+
+
+  public int size() {
+    return rowQueue.size();
+  }
+
+
+  public boolean isEmpty() {
+    return rowQueue.isEmpty();
+  }
+
+  // Unlike push queries that run forever until someone deliberately kills it, pull queries have an
+  // ending.  When they've reached their end, this is expected to be called.  Also, if the system
+  // wants to end pull queries prematurely, this should also be called.
+  public void close() {
+    closed = true;
+    // Unlike limits based on a number of rows which can be checked and possibly triggered after
+    // every queuing of a row, pull queries just declare they've reached their limit when close is
+    // called.
```

Comment:
what happens when we implement a `LIMIT` clause on pull queries (such as table scans)? Would we want to call `limitReached` twice or would that just be a physical plan node that just calls close on the underlying queue?

Reply:
Hmm... in SQL the LIMIT clause is deterministic usually, even if there's no ORDER BY (in which case, it would follow the row ordering).

In push / persistent queries this is not hard to maintain, as we naturally follow the offset ordering. But in pull query does this mean we may then return different results if the same query with LIMIT and no ORDER BY is issued multiple times?